### PR TITLE
add search filter

### LINKED ISSUE #9: 9. As a user, I want to filter my shopping list to make it easier to locate an item in the list.
AC:

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should _not_ only match from the start of the string)

### PR DESCRIPTION
**Pull request by sicunchen**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

Inside List.js, we added a search input and a handleSearch event handler to narrow down the list based on the user input. We also solved a critical bug that caused infinite render of the list in ItemsContext.js. We converted the token from a static value from localStorage into a dynamic value through the useTokenHook and therefore the useEffect could be updated properly.
## Related Issue

closes #9 
## Acceptance Criteria

* Display a text field above the top of the shopping list
* As the user types into the field, the list should narrow to display only items that contain the text the user entered in the filter field
* When the field has text in it, the user should be able to tap a UI element (e.g., with an "X" button next to the field) to clear the field
* The filter text should match any part of the item name (i.e. it should not only match from the start of the string)

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

### Before

Screen Shot 2020-09-03 at 20 41 30

### After

Screen Shot 2020-09-03 at 20 41 41

### CODE REVIEW

**Review by MabelM1 — APPROVED** [review]

File: src/ItemsContext.js
Comment (lacrivella) [inline src/ItemsContext.js:13]: Super grateful for you fixing our projects bug! I didn't even know where to start with this issue, but y'all, @MabelM1, and @skillitzimberg were able to figure it out. 💜

File: src/List.js
Comment (lacrivella) [inline src/List.js:39]: Just from a UX/UI perspective, when a user filters through their list and the item doesn't exist, we should maybe have a different message like 'item not found' or 'item does not exist, please add to your list'. I still think we should have the 'your list is empty, please add an item' for when the list is completely empty.

This isn't a critical thing, but maybe a stretch goal for the future.

**Review by lacrivella — APPROVED** [review]
Awesome job! I left two notes, but they're nothing that require changes before approval.

**Review by skillitzimberg — APPROVED** [review]
Awesome job teaming up and fixing that re-render bug! I think @lacrivella is right about the messaging around an empty list vs no items being found on filter. Maybe we can find an issue that would makes sense to add that to . . . ? Or maybe you all can just decide which team will make that fix on the next round of PRs.

File: src/List.js
Comment (yvesgurcan) [reply src/List.js:39]: Absolutely, Let’s have a different message for the case where there are no items in the list and the case where no items match the search term

**Review by yvesgurcan — APPROVED** [review]
Nice and simple. Well done.

**Pull request merged**
